HouseRules_Essentials/Rules/MonsterDeckOverriddenRule.cs

```diff
@@ -0,0 +1,229 @@
+                    BoardPieceId = deckItemConfig.Key,
+                    enemyWeight = EnemyWeight.Light,
+                    isRedrawEnabled = false,
+                };
+                if (deckItemConfig.Value == 0)
+                {
+                    deckItem.isRedrawEnabled = true;
+                }
+
+                mySubDeck.Add(deckItem);
+                for (int i = 0; i < Math.Max(0, deckItemConfig.Value - 1); i++)
+                {
+                    mySubDeck.Add(deckItem);
+                }
+            }
+
+            return mySubDeck;
+        }
+
+        private static bool AIDirectorDeckConstructor_ConstructMonsterDeck_Prefix(ref MonsterDeck __result, ISpawnCategoryProvider spawnCategoryProvider, int floorIndex, IRnd rng, LevelSequence.GameType gameType)
+        {
+            if (!_isActivated)
+            {
+                return true;
+            }
+
+            List<MonsterDeck.MonsterDeckEntry> standardDeck;
+            List<MonsterDeck.MonsterDeckEntry> spikeDeck;
+            MonsterDeck.MonsterDeckEntry keyHolder;
+            if (floorIndex == 1)
+            {
+                standardDeck = CreateSubDeck(_globalAdjustments.EntranceDeckFloor1);
+                spikeDeck = CreateSubDeck(_globalAdjustments.ExitDeckFloor1);
+                keyHolder = new MonsterDeck.MonsterDeckEntry() { BoardPieceId = _globalAdjustments.KeyHolderFloor1, enemyWeight = EnemyWeight.Light, isRedrawEnabled = false };
+            }
+            else
+            {
+                standardDeck = CreateSubDeck(_globalAdjustments.EntranceDeckFloor2);
+                spikeDeck = CreateSubDeck(_globalAdjustments.ExitDeckFloor2);
+                keyHolder = new MonsterDeck.MonsterDeckEntry() { BoardPieceId = _globalAdjustments.KeyHolderFloor2, enemyWeight = EnemyWeight.Light, isRedrawEnabled = false };
+            }
+
+            List<MonsterDeck.MonsterDeckEntry> bossDeck = CreateSubDeck(_globalAdjustments.BossDeck);
+
+            MonsterDeck monsterDeck = new MonsterDeck
+            {
+                monsterDeckStandard = standardDeck,
+                monsterDeckSpike = spikeDeck,
+                monsterDeckBoss = bossDeck,
+                Keyholder = keyHolder,
+            };
+            monsterDeck.ShuffleSubDecks(rng);
+            __result = monsterDeck;
+            return false; // We returned an user-adjusted config.
+        }
+
+        private static bool AIDirectorDeckConstructor_AddPreFillUnitsToDeck_Prefix(List<MonsterDeck.MonsterDeckEntry> monsterDeckList, ISpawnCategoryProvider spawnCategoryProvider)
+        {
+            if (!_isActivated)
+            {
+                return true;
+            }
+
+            return false; // We did all that we needed to do, and that was nothing.
+        }
+
+        private static bool AIDirectorController2_SpawnBossAndMinions_Prefix(ref AIDirectorController2 __instance, ref AIDirectorContext context, IRnd rng, ref TransientBoardState boardState)
```

Comment:
The contents of this method, and the params, look like the function `PickSpawnZoneForKeyholder()` and not `SpawnBossAndMinions`.  Is this intentional?

Reply:
Follow-up as I think I understand now:

Are we recreating `PickSpawnZoneForKeyholder()` here and then using the result directly in order to shorten `SpawnBossAndMinions()`?

If so, then another question:

Are we doing anything different than `PickSpawnZoneForKeyholder()` already does (nothing different jumps out at me).  If not, then we might want to consider calling `PickSpawnZoneForKeyholder()` instead of rewriting it's logic.

Reply:
I'd been stalled for over a week trying to get the same result out using Traverse.Create, but was getting nowhere. No matter what I tried, the keyHolderPosition always came back with `x=0`, `y=0`

```
            var spawnZone = Traverse.Create(pickSpawnZoneForKeyholderResult).Field<SpawnZone>("spawnZone").Value;
            var keyHolderPosition = Traverse.Create(pickSpawnZoneForKeyholderResult).Field<IntPoint2D>("keyHolderPosition").Value;

```

If you've got time to try and make this work, there's an earlier commit that was attempting it.. https://github.com/orendain/DemeoMods/blob/80606e71876970c345bec6b9211526a85d561844/HouseRules_Essentials/Rules/MonsterDeckOverriddenRule.cs#L180

After being stalled for too long, I finally worked around the issue by moving the required code from pickSpawnZoneForKeyholder into the patch here.